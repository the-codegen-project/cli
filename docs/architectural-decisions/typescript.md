# Architectural decisions for TypeScript
All decision worth mentioning about TypeScript can be found here, simply to keep track.

### 21.08.2024

1. All functions in channel protocols are to be designed arrow function as it's more versatile when combining it together with other code sections.
