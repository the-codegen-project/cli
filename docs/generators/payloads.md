---
sidebar_position: 99
---

# 🐔 Payloads

```js
export default {
  ...,
  generators: [
    {
      preset: 'payloads',
      outputPath: './src/payloads',
      serializationType: 'json', 
  	  language: 'typescript',
    }
  ]
};
```

`payloads` preset is for generating models that represent typed models that can be serialized into message payloads for communication use-cases.

This is supported through the following inputs: `asyncapi`, `openapi`

It supports the following languages; [`typescript`](#typescript)

## Languages
Each language has a set of constraints which means that some typed model types are either supported or not, or it might just be the code generation library that does not yet support it.

|  | Circular models | Enums | Tuples | Arrays | Nested Arrays | Dictionaries | Json Serialization | Validation |
|---|---|---|---|---|---|---|---|---|
| **TypeScript** | ✔️ | ✔️ | ✔️ | ✔️ | ✔️ | ✔️ | ✔️ | ✔️ |

### TypeScript

Dependencies: 
- If validation enabled, [ajv](https://ajv.js.org/guide/getting-started.html): ^8.17.1

#### Validation
Each generated class includes built-in JSON Schema validation capabilities through two static methods:

- `validate`: Validates data against the schema. Use this method when you want to validate data.

```typescript
// Example
const result = UserSignedUp.validate({ data: userData });
if (!result.valid) {
  console.error('Validation errors:', result.errors);
}
```

- `createValidator`: Creates a reusable validator function. Use this when you need to validate multiple instances of the same type and want to avoid recreating the validator each time.

```typescript
// Example
const validator = UserSignedUp.createValidator();
const result = UserSignedUp.validate({ data: userData, ajvValidatorFunction: validator });
if (!result.valid) {
  console.error('Validation errors:', result.errors);
}
```

Both methods support custom Ajv instances and options for advanced validation scenarios.
