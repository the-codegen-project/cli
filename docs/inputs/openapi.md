---
sidebar_position: 99
---

# OpenAPI

Input support; `openapi`

- OpenAPI 3.0.x
- OpenAPI 3.1.x
- OpenAPI 2.0.0 (Swagger)

| **Presets** | OpenAPI | 
|---|---|
| [`payloads`](../generators/payloads.md) | ✅ |
| [`parameters`](../generators/parameters.md) | ✅ |
| [`headers`](../generators/headers.md) | ✅ |
| [`types`](../generators/types.md) | ✅ |
| [`channels`](../generators/channels.md) | ❌ |
| [`client`](../generators/client.md) | ❌ |
| [`custom`](../generators/custom.md) | ✅ |

## Basic Usage

### Configuration

Create a configuration file that specifies OpenAPI as the input type:

```json
{
  "inputType": "openapi",
  "inputPath": "./api/openapi.yaml",
  "language": "typescript",
  "generators": [ ... ]
}
```

## Advanced Features

### External References

The OpenAPI parser automatically resolves external `$ref` references:

```yaml
components:
  schemas:
    Pet:
      $ref: './schemas/pet.yaml#/Pet'
    User:
      $ref: 'https://api.example.com/schemas/user.json#/User'
```

### OpenAPI 3.1 Features

Full support for OpenAPI 3.1 features including:

- JSON Schema 2020-12 compatibility
- `const` keyword
- `if`/`then`/`else` conditionals
- Enhanced `examples` support

### Validation and Error Handling

The parser provides detailed validation errors:

```typescript
// If validation fails, you'll get detailed error information
try {
  const document = await loadOpenapi(context);
} catch (error) {
  console.error('OpenAPI validation failed:', error.message);
  // Error message includes line numbers and specific validation issues
}
```

## Examples

### REST API Client Generation

Generate a complete TypeScript client for your REST API:

```json
{
  "inputType": "openapi",
  "inputPath": "./api/openapi.yaml",
  "language": "typescript",
  "generators": [ ]
}
```

## Best Practices

1. **Schema Organization**: Use `$ref` to organize complex schemas into separate files
2. **Validation**: Always validate your OpenAPI documents before generation
3. **Versioning**: Include version information in your API specifications
4. **Documentation**: Use `description` fields extensively for better generated code
5. **Examples**: Provide examples in your schemas for better understanding

## Troubleshooting

### Common Issues

1. **Invalid $ref**: Ensure all referenced files exist and are accessible
2. **Schema Validation**: Check that your OpenAPI document follows the specification
3. **File Format**: Verify that YAML/JSON syntax is correct
4. **Circular References**: Avoid circular `$ref` dependencies

## FAQ

### Can I use both OpenAPI and AsyncAPI in the same project?

Yes! You can have separate configuration files for each input type and generate code to different output directories.

### What's the difference between OpenAPI 3.0 and 3.1?

OpenAPI 3.1 is fully compatible with JSON Schema 2020-12 and includes additional features like `const`, conditional schemas, and enhanced examples support.

### How do I handle authentication in generated clients?

Define security schemes in your OpenAPI document, and the generated client code will include appropriate authentication handling.

### Can I customize the generated code?

Yes, use the custom generator preset to create your own templates and generation logic. 