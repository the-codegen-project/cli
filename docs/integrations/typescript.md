---
sidebar_position: 99
---

# TypeScript Project

Simple TypeScript library that use AsyncAPI as input to generate payload models that is serialized and printed in the console.

[See the project in action ðŸŽ¬](https://github.com/the-codegen-project/cli/tree/main/examples/typescript-library)