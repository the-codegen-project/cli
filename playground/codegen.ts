import { TheCodegenConfiguration } from "@the-codegen-project/cli";
const config: TheCodegenConfiguration = {
  inputType: "asyncapi",
  inputPath: "./schema.json",
  language: "typescript",
  generators: [
    {
      preset: "channels",
      outputPath: "./__gen__/channels",
      protocols: ["nats"],
    },
    {
      preset: "payloads",
      outputPath: "./__gen__/payloads",
      serializationType: "json",
      model: "class",
      enum: "enum",
      map: "record",
      moduleSystem: "esm",
      useForJavaScript: true,
      rawPropertyNames: false,
    },
    {
      preset: "parameters",
      outputPath: "./__gen__/parameters",
      serializationType: "json",
    },
  ],
};
export default config;
