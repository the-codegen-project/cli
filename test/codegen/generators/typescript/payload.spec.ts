import path from "node:path";
import { defaultTypeScriptPayloadGenerator, generateTypescriptPayload } from "../../../../src/codegen/generators";
import { loadAsyncapiDocument } from "../../../../src/codegen/inputs/asyncapi";
import { loadOpenapiDocument } from "../../../../src/codegen/inputs/openapi";
import { safeStringify } from "../../../../src/codegen/generators/typescript/payloads";

describe('payloads', () => {
  describe('typescript', () => {
    it('should work with basic AsyncAPI inputs', async () => {
      const parsedAsyncAPIDocument = await loadAsyncapiDocument(path.resolve(__dirname, '../../../configs/payload.yaml'));
      
      const renderedContent = await generateTypescriptPayload({
        generator: {
          ...defaultTypeScriptPayloadGenerator,
          outputPath: path.resolve(__dirname, './output')
        },
        inputType: 'asyncapi',
        asyncapiDocument: parsedAsyncAPIDocument,
        dependencyOutputs: { }
      });
      expect(renderedContent.channelModels['union'].messageModel.result).toMatchSnapshot();
      expect(renderedContent.channelModels['simple'].messageModel.result).toMatchSnapshot();
    });
    it('should not render validation functions', async () => {
      const parsedAsyncAPIDocument = await loadAsyncapiDocument(path.resolve(__dirname, '../../../configs/payload.yaml'));
      
      const renderedContent = await generateTypescriptPayload({
        generator: {
          ...defaultTypeScriptPayloadGenerator,
          includeValidation: false,
          outputPath: path.resolve(__dirname, './output')
        },
        inputType: 'asyncapi',
        asyncapiDocument: parsedAsyncAPIDocument,
        dependencyOutputs: { }
      });
      expect(renderedContent.channelModels['union'].messageModel.result).toMatchSnapshot();
      expect(renderedContent.channelModels['simple'].messageModel.result).toMatchSnapshot();
    });
    it('should work with no channels', async () => {
      const parsedAsyncAPIDocument = await loadAsyncapiDocument(path.resolve(__dirname, '../../../configs/payload-no-channels.yaml'));
      
      const renderedContent = await generateTypescriptPayload({
        generator: {
          ...defaultTypeScriptPayloadGenerator,
          outputPath: path.resolve(__dirname, './output')
        },
        inputType: 'asyncapi',
        asyncapiDocument: parsedAsyncAPIDocument,
        dependencyOutputs: { }
      });
      expect(renderedContent.otherModels.length).toEqual(1);
      expect(renderedContent.otherModels[0].messageModel.result).toMatchSnapshot();
    });
    it('should get correct model names', async () => {
      const parsedAsyncAPIDocument = await loadAsyncapiDocument(path.resolve(__dirname, '../../../configs/payload-complex.json'));
      
      const renderedContent = await generateTypescriptPayload({
        generator: {
          ...defaultTypeScriptPayloadGenerator,
          outputPath: path.resolve(__dirname, './output')
        },
        inputType: 'asyncapi',
        asyncapiDocument: parsedAsyncAPIDocument,
        dependencyOutputs: { }
      });
      const payloadNames = renderedContent.otherModels.map((value) => value.messageModel.modelName);
      expect(payloadNames.includes('ComplexMessage')).toEqual(true);
      expect(payloadNames.includes('OpenApiGenerator')).toEqual(true);
      expect(payloadNames.includes('AsyncApiGenerator')).toEqual(true);
      expect(payloadNames.includes('DocumentContext')).toEqual(true);
      expect(payloadNames.includes('Simple')).toEqual(true);
      expect(payloadNames.includes('NpmRelease')).toEqual(true);
      expect(payloadNames.includes('GeneratorTypes')).toEqual(true);
      expect(payloadNames.includes('TypescriptNatsSettings')).toEqual(true);
      expect(payloadNames.includes('OpenapiTypescriptFetch')).toEqual(true);
    });
    
    it('should work with basic OpenAPI 2.0 inputs', async () => {
      const parsedOpenAPIDocument = await loadOpenapiDocument(path.resolve(__dirname, '../../../configs/openapi-2.json'));
      
      const renderedContent = await generateTypescriptPayload({
        generator: {
          ...defaultTypeScriptPayloadGenerator,
          outputPath: path.resolve(__dirname, './output')
        },
        inputType: 'openapi',
        openapiDocument: parsedOpenAPIDocument,
        dependencyOutputs: { }
      });
      const modelsCreated = Object.entries(renderedContent.operationModels).map(([key, value]) => ({key, value: value.messageModel.modelName}));
      expect(modelsCreated).toEqual([
        {
          key: "addPet",
          value: "AddPetRequest",
        },
        {
          key: "updatePet",
          value: "UpdatePetRequest",
        },
        {
          key: "placeOrder",
          value: "PlaceOrderRequest",
        },
        {
          key: "createUsersWithListInput",
          value: "CreateUsersWithListInputRequest",
        },
        {
          key: "updateUser",
          value: "UpdateUserRequest",
        },
        {
          key: "createUsersWithArrayInput",
          value: "CreateUsersWithArrayInputRequest",
        },
        {
          key: "createUser",
          value: "CreateUserRequest",
        },
        {
          key: "uploadFile_Response",
          value: "UploadFileResponse_200",
        },
        {
          key: "findPetsByStatus_Response",
          value: "FindPetsByStatusResponse_200",
        },
        {
          key: "findPetsByTags_Response",
          value: "FindPetsByTagsResponse_200",
        },
        {
          key: "getPetById_Response",
          value: "GetPetByIdResponse_200",
        },
        {
          key: "getInventory_Response",
          value: "GetInventoryResponse_200",
        },
        {
          key: "placeOrder_Response",
          value: "PlaceOrderResponse_200",
        },
        {
          key: "getOrderById_Response",
          value: "GetOrderByIdResponse_200",
        },
        {
          key: "getUserByName_Response",
          value: "GetUserByNameResponse_200",
        },
        {
          key: "loginUser_Response",
          value: "LoginUserResponse_200",
        },
      ]);
    });
    it('should work with basic OpenAPI 3.0 inputs', async () => {
      const parsedOpenAPIDocument = await loadOpenapiDocument(path.resolve(__dirname, '../../../configs/openapi-3.json'));
      
      const renderedContent = await generateTypescriptPayload({
        generator: {
          ...defaultTypeScriptPayloadGenerator,
          outputPath: path.resolve(__dirname, './output')
        },
        inputType: 'openapi',
        openapiDocument: parsedOpenAPIDocument,
        dependencyOutputs: { }
      });
      const modelsCreated = Object.entries(renderedContent.operationModels).map(([key, value]) => ({key, value: value.messageModel.modelName}));
      expect(modelsCreated).toEqual([
        {
          key: "addPet",
          value: "APet",
        },
        {
          key: "updatePet",
          value: "APet",
        },
        {
          key: "placeOrder",
          value: "PetOrder",
        },
        {
          key: "createUser",
          value: "AUser",
        },
        {
          key: "createUsersWithArrayInput",
          value: "CreateUsersWithArrayInputRequest",
        },
        {
          key: "createUsersWithListInput",
          value: "CreateUsersWithListInputRequest",
        },
        {
          key: "addPet_Response",
          value: "APet",
        },
        {
          key: "updatePet_Response",
          value: "APet",
        },
        {
          key: "findPetsByStatus_Response",
          value: "FindPetsByStatusResponse_200",
        },
        {
          key: "findPetsByTags_Response",
          value: "FindPetsByTagsResponse_200",
        },
        {
          key: "getPetById_Response",
          value: "APet",
        },
        {
          key: "uploadFile_Response",
          value: "AnUploadedResponse",
        },
        {
          key: "getInventory_Response",
          value: "GetInventoryResponse_200",
        },
        {
          key: "placeOrder_Response",
          value: "PetOrder",
        },
        {
          key: "getOrderById_Response",
          value: "PetOrder",
        },
        {
          key: "loginUser_Response",
          value: "LoginUserResponse_200",
        },
        {
          key: "getUserByName_Response",
          value: "AUser",
        },
      ]);
    });
    it('should work with basic OpenAPI 3.1 inputs', async () => {
      const parsedOpenAPIDocument = await loadOpenapiDocument(path.resolve(__dirname, '../../../configs/openapi-3_1.json'));
      
      const renderedContent = await generateTypescriptPayload({
        generator: {
          ...defaultTypeScriptPayloadGenerator,
          outputPath: path.resolve(__dirname, './output')
        },
        inputType: 'openapi',
        openapiDocument: parsedOpenAPIDocument,
        dependencyOutputs: { }
      });
      const modelsCreated = Object.entries(renderedContent.operationModels).map(([key, value]) => ({key, value: value.messageModel.modelName}));
      expect(modelsCreated).toEqual([
        {
          key: "addPet",
          value: "APet",
        },
        {
          key: "updatePet",
          value: "APet",
        },
        {
          key: "placeOrder",
          value: "PetOrder",
        },
        {
          key: "createUser",
          value: "AUser",
        },
        {
          key: "createUsersWithArrayInput",
          value: "CreateUsersWithArrayInputRequest",
        },
        {
          key: "createUsersWithListInput",
          value: "CreateUsersWithListInputRequest",
        },
        {
          key: "addPet_Response",
          value: "APet",
        },
        {
          key: "updatePet_Response",
          value: "APet",
        },
        {
          key: "findPetsByStatus_Response",
          value: "FindPetsByStatusResponse_200",
        },
        {
          key: "findPetsByTags_Response",
          value: "FindPetsByTagsResponse_200",
        },
        {
          key: "getPetById_Response",
          value: "APet",
        },
        {
          key: "uploadFile_Response",
          value: "AnUploadedResponse",
        },
        {
          key: "getInventory_Response",
          value: "GetInventoryResponse_200",
        },
        {
          key: "placeOrder_Response",
          value: "PetOrder",
        },
        {
          key: "getOrderById_Response",
          value: "PetOrder",
        },
        {
          key: "loginUser_Response",
          value: "LoginUserResponse_200",
        },
        {
          key: "getUserByName_Response",
          value: "AUser",
        },
      ]);
    });
    describe('safeStringify', () => {
      it('should stringify basic primitive values', () => {
        expect(safeStringify('hello')).toBe('"hello"');
        expect(safeStringify(42)).toBe('42');
        expect(safeStringify(true)).toBe('true');
        expect(safeStringify(false)).toBe('false');
        expect(safeStringify(null)).toBe('null');
      });

      it('should stringify simple objects', () => {
        const obj = { name: 'test', value: 123 };
        expect(safeStringify(obj)).toBe('{"name":"test","value":123}');
      });

      it('should stringify arrays', () => {
        const arr = [1, 'two', true, null];
        expect(safeStringify(arr)).toBe('[1,"two",true,null]');
      });

      it('should remove x-modelina extension properties', () => {
        const obj = {
          name: 'test',
          'x-modelina-type': 'string',
          'x-modelina-format': 'email',
          value: 123
        };
        expect(safeStringify(obj)).toBe('{"name":"test","value":123}');
      });

      it('should remove x-the-codegen-project extension properties', () => {
        const obj = {
          name: 'test',
          'x-the-codegen-project-version': '1.0.0',
          'x-the-codegen-project-config': { setting: true },
          value: 123
        };
        expect(safeStringify(obj)).toBe('{"name":"test","value":123}');
      });

      it('should handle circular references by replacing with true', () => {
        const obj: any = { name: 'test' };
        obj.self = obj;
        const result = safeStringify(obj);
        expect(result).toBe('{"name":"test","self":true}');
      });

      it('should handle nested circular references by cutting off after repetitions', () => {
        const parent: any = { name: 'parent' };
        const child: any = { name: 'child', parent };
        parent.child = child;
        
        const result = safeStringify(parent);
        // Should allow some repetition but eventually cut off with true
        expect(result).toContain('"name":"parent"');
        expect(result).toContain('"name":"child"');
        expect(result).toContain('true');
        // Should not be infinite
        expect(result.length).toBeLessThan(1000);
      });
      
      it('should handle multiple references to the same object', () => {
        const parent: any = { name: 'parent' };
        const root: any = { name: 'child', parent, parent_old: parent };
        
        const result = safeStringify(root);
        expect(result).toBe('{"name":"child","parent":{"name":"parent"},"parent_old":{"name":"parent"}}');
      });

      it('should replace functions with true', () => {
        const obj = {
          name: 'test',
          method() { return 'hello'; },
          arrow: () => 'world',
          value: 123
        };
        expect(safeStringify(obj)).toBe('{"name":"test","method":true,"arrow":true,"value":123}');
      });

      it('should handle deeply nested objects within recursion limit', () => {
        let nested: any = { value: 'deep' };
        for (let i = 0; i < 10; i++) {
          nested = { level: i, nested };
        }
        
        const result = safeStringify(nested);
        expect(result).toContain('"value":"deep"');
        expect(result).toContain('"level":0');
      });

      it('should handle objects beyond recursion limit by replacing with true', () => {
        let nested: any = { value: 'deep' };
        // Create a very deep nesting that exceeds the 255 limit
        for (let i = 0; i < 300; i++) {
          nested = { level: i, nested };
        }
        
        const result = safeStringify(nested);
        // Should contain true for deeply nested parts
        expect(result).toContain('true');
      });

      it('should handle mixed extension properties and regular properties', () => {
        const obj = {
          name: 'test',
          'x-modelina-discriminator': 'type',
          description: 'A test object',
          'x-the-codegen-project-id': 'test-id',
          properties: {
            field1: 'value1',
            'x-modelina-required': true,
            field2: 'value2'
          }
        };
        
        const result = safeStringify(obj);
        const parsed = JSON.parse(result);
        
        expect(parsed.name).toBe('test');
        expect(parsed.description).toBe('A test object');
        expect(parsed.properties.field1).toBe('value1');
        expect(parsed.properties.field2).toBe('value2');
        expect(parsed['x-modelina-discriminator']).toBeUndefined();
        expect(parsed['x-the-codegen-project-id']).toBeUndefined();
        expect(parsed.properties['x-modelina-required']).toBeUndefined();
      });

      it('should handle undefined values', () => {
        const obj = {
          name: 'test',
          undefinedValue: undefined,
          value: 123
        };
        expect(safeStringify(obj)).toBe('{"name":"test","value":123}');
      });

      it('should handle empty objects and arrays', () => {
        expect(safeStringify({})).toBe('{}');
        expect(safeStringify([])).toBe('[]');
      });

      it('should handle complex nested structures with extension properties', () => {
        const complex = {
          schema: {
            type: 'object',
            'x-modelina-type': 'CustomType',
            properties: {
              name: {
                type: 'string',
                'x-the-codegen-project-validation': 'required'
              },
              items: {
                type: 'array',
                'x-modelina-items': 'string',
                items: {
                  type: 'string'
                }
              }
            }
          },
          'x-the-codegen-project-version': '2.0.0'
        };

        const result = safeStringify(complex);
        const parsed = JSON.parse(result);
        
        expect(parsed.schema.type).toBe('object');
        expect(parsed.schema.properties.name.type).toBe('string');
        expect(parsed.schema.properties.items.type).toBe('array');
        expect(parsed.schema.properties.items.items.type).toBe('string');
        
        // Extension properties should be removed
        expect(parsed.schema['x-modelina-type']).toBeUndefined();
        expect(parsed.schema.properties.name['x-the-codegen-project-validation']).toBeUndefined();
        expect(parsed.schema.properties.items['x-modelina-items']).toBeUndefined();
        expect(parsed['x-the-codegen-project-version']).toBeUndefined();
      });
    });
  });
});
