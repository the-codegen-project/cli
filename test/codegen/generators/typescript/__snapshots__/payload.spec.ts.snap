// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`payloads typescript should not render validation functions 1`] = `
"import {SimpleObject} from './SimpleObject';
type UnionPayload = SimpleObject | boolean | string;

export function unmarshal(json: any): UnionPayload {
  if(typeof json === 'object') {
    if(json.type === 'SimpleObject') {
  return SimpleObject.unmarshal(json);
  }
  }
  return JSON.parse(json);
}
export function marshal(payload: UnionPayload) {
  if(payload instanceof SimpleObject) {
return payload.marshal();
}


  return JSON.stringify(payload);
}
export { UnionPayload };"
`;

exports[`payloads typescript should not render validation functions 2`] = `
"
class SimpleObject2 {
  private _displayName?: string;
  private _email?: string;
  private _additionalProperties?: Record<string, any>;

  constructor(input: {
    displayName?: string,
    email?: string,
    additionalProperties?: Record<string, any>,
  }) {
    this._displayName = input.displayName;
    this._email = input.email;
    this._additionalProperties = input.additionalProperties;
  }

  get displayName(): string | undefined { return this._displayName; }
  set displayName(displayName: string | undefined) { this._displayName = displayName; }

  get email(): string | undefined { return this._email; }
  set email(email: string | undefined) { this._email = email; }

  get additionalProperties(): Record<string, any> | undefined { return this._additionalProperties; }
  set additionalProperties(additionalProperties: Record<string, any> | undefined) { this._additionalProperties = additionalProperties; }

  public marshal() : string {
    let json = '{'
    if(this.displayName !== undefined) {
      json += \`"displayName": \${typeof this.displayName === 'number' || typeof this.displayName === 'boolean' ? this.displayName : JSON.stringify(this.displayName)},\`;
    }
    if(this.email !== undefined) {
      json += \`"email": \${typeof this.email === 'number' || typeof this.email === 'boolean' ? this.email : JSON.stringify(this.email)},\`;
    }
    if(this.additionalProperties !== undefined) { 
      for (const [key, value] of this.additionalProperties.entries()) {
        //Only unwrap those that are not already a property in the JSON object
        if(["displayName","email","additionalProperties"].includes(String(key))) continue;
        json += \`"\${key}": \${typeof value === 'number' || typeof value === 'boolean' ? value : JSON.stringify(value)},\`;
      }
    }
    //Remove potential last comma 
    return \`\${json.charAt(json.length-1) === ',' ? json.slice(0, json.length-1) : json}}\`;
  }

  public static unmarshal(json: string | object): SimpleObject2 {
    const obj = typeof json === "object" ? json : JSON.parse(json);
    const instance = new SimpleObject2({} as any);

    if (obj["displayName"] !== undefined) {
      instance.displayName = obj["displayName"];
    }
    if (obj["email"] !== undefined) {
      instance.email = obj["email"];
    }
  
    instance.additionalProperties = new Map();
    const propsToCheck = Object.entries(obj).filter((([key,]) => {return !["displayName","email","additionalProperties"].includes(key);}));
    for (const [key, value] of propsToCheck) {
      instance.additionalProperties.set(key, value as any);
    }
    return instance;
  }
}
export { SimpleObject2 };"
`;

exports[`payloads typescript should work with basic AsyncAPI inputs 1`] = `
"import {SimpleObject} from './SimpleObject';
type UnionPayload = SimpleObject | boolean | string;

export function unmarshal(json: any): UnionPayload {
  if(typeof json === 'object') {
    if(json.type === 'SimpleObject') {
  return SimpleObject.unmarshal(json);
  }
  }
  return JSON.parse(json);
}
export function marshal(payload: UnionPayload) {
  if(payload instanceof SimpleObject) {
return payload.marshal();
}


  return JSON.stringify(payload);
}
export { UnionPayload };"
`;

exports[`payloads typescript should work with basic AsyncAPI inputs 2`] = `
"import {Ajv, Options as AjvOptions, ErrorObject, ValidateFunction} from 'ajv';
import addFormats from 'ajv-formats';
class SimpleObject2 {
  private _displayName?: string;
  private _email?: string;
  private _additionalProperties?: Record<string, any>;

  constructor(input: {
    displayName?: string,
    email?: string,
    additionalProperties?: Record<string, any>,
  }) {
    this._displayName = input.displayName;
    this._email = input.email;
    this._additionalProperties = input.additionalProperties;
  }

  get displayName(): string | undefined { return this._displayName; }
  set displayName(displayName: string | undefined) { this._displayName = displayName; }

  get email(): string | undefined { return this._email; }
  set email(email: string | undefined) { this._email = email; }

  get additionalProperties(): Record<string, any> | undefined { return this._additionalProperties; }
  set additionalProperties(additionalProperties: Record<string, any> | undefined) { this._additionalProperties = additionalProperties; }

  public marshal() : string {
    let json = '{'
    if(this.displayName !== undefined) {
      json += \`"displayName": \${typeof this.displayName === 'number' || typeof this.displayName === 'boolean' ? this.displayName : JSON.stringify(this.displayName)},\`;
    }
    if(this.email !== undefined) {
      json += \`"email": \${typeof this.email === 'number' || typeof this.email === 'boolean' ? this.email : JSON.stringify(this.email)},\`;
    }
    if(this.additionalProperties !== undefined) { 
      for (const [key, value] of this.additionalProperties.entries()) {
        //Only unwrap those that are not already a property in the JSON object
        if(["displayName","email","additionalProperties"].includes(String(key))) continue;
        json += \`"\${key}": \${typeof value === 'number' || typeof value === 'boolean' ? value : JSON.stringify(value)},\`;
      }
    }
    //Remove potential last comma 
    return \`\${json.charAt(json.length-1) === ',' ? json.slice(0, json.length-1) : json}}\`;
  }

  public static unmarshal(json: string | object): SimpleObject2 {
    const obj = typeof json === "object" ? json : JSON.parse(json);
    const instance = new SimpleObject2({} as any);

    if (obj["displayName"] !== undefined) {
      instance.displayName = obj["displayName"];
    }
    if (obj["email"] !== undefined) {
      instance.email = obj["email"];
    }
  
    instance.additionalProperties = new Map();
    const propsToCheck = Object.entries(obj).filter((([key,]) => {return !["displayName","email","additionalProperties"].includes(key);}));
    for (const [key, value] of propsToCheck) {
      instance.additionalProperties.set(key, value as any);
    }
    return instance;
  }
  public static theCodeGenSchema = {"type":"object","$schema":"http://json-schema.org/draft-07/schema","properties":{"displayName":{"type":"string","description":"Name of the user"},"email":{"type":"string","format":"email","description":"Email of the user"}},"$id":"SimpleObject2"};
  public static validate(context?: {data: any, ajvValidatorFunction?: ValidateFunction, ajvInstance?: Ajv, ajvOptions?: AjvOptions}): { valid: boolean; errors?: ErrorObject[]; } {
    const {data, ajvValidatorFunction} = context ?? {};
    const parsedData = typeof data === 'string' ? JSON.parse(data) : data;
    const validate = ajvValidatorFunction ?? this.createValidator(context)
    return {
      valid: validate(parsedData),
      errors: validate.errors ?? undefined,
    };
  }
  public static createValidator(context?: {ajvInstance?: Ajv, ajvOptions?: AjvOptions}): ValidateFunction {
    const {ajvInstance} = {...context ?? {}, ajvInstance: new Ajv(context?.ajvOptions ?? {})};
    addFormats(ajvInstance);
    const validate = ajvInstance.compile(this.theCodeGenSchema);
    return validate;
  }

}
export { SimpleObject2 };"
`;

exports[`payloads typescript should work with no channels 1`] = `
"import {Ajv, Options as AjvOptions, ErrorObject, ValidateFunction} from 'ajv';
import addFormats from 'ajv-formats';
class AnonymousSchema_1 {
  private _type?: 'SimpleObject' = 'SimpleObject';
  private _displayName?: string;
  private _email?: string;
  private _additionalProperties?: Record<string, any>;

  constructor(input: {
    displayName?: string,
    email?: string,
    additionalProperties?: Record<string, any>,
  }) {
    this._displayName = input.displayName;
    this._email = input.email;
    this._additionalProperties = input.additionalProperties;
  }

  get type(): 'SimpleObject' | undefined { return this._type; }

  get displayName(): string | undefined { return this._displayName; }
  set displayName(displayName: string | undefined) { this._displayName = displayName; }

  get email(): string | undefined { return this._email; }
  set email(email: string | undefined) { this._email = email; }

  get additionalProperties(): Record<string, any> | undefined { return this._additionalProperties; }
  set additionalProperties(additionalProperties: Record<string, any> | undefined) { this._additionalProperties = additionalProperties; }

  public marshal() : string {
    let json = '{'
    if(this.type !== undefined) {
      json += \`"type": \${typeof this.type === 'number' || typeof this.type === 'boolean' ? this.type : JSON.stringify(this.type)},\`;
    }
    if(this.displayName !== undefined) {
      json += \`"displayName": \${typeof this.displayName === 'number' || typeof this.displayName === 'boolean' ? this.displayName : JSON.stringify(this.displayName)},\`;
    }
    if(this.email !== undefined) {
      json += \`"email": \${typeof this.email === 'number' || typeof this.email === 'boolean' ? this.email : JSON.stringify(this.email)},\`;
    }
    if(this.additionalProperties !== undefined) { 
      for (const [key, value] of this.additionalProperties.entries()) {
        //Only unwrap those that are not already a property in the JSON object
        if(["type","displayName","email","additionalProperties"].includes(String(key))) continue;
        json += \`"\${key}": \${typeof value === 'number' || typeof value === 'boolean' ? value : JSON.stringify(value)},\`;
      }
    }
    //Remove potential last comma 
    return \`\${json.charAt(json.length-1) === ',' ? json.slice(0, json.length-1) : json}}\`;
  }

  public static unmarshal(json: string | object): AnonymousSchema_1 {
    const obj = typeof json === "object" ? json : JSON.parse(json);
    const instance = new AnonymousSchema_1({} as any);

    if (obj["displayName"] !== undefined) {
      instance.displayName = obj["displayName"];
    }
    if (obj["email"] !== undefined) {
      instance.email = obj["email"];
    }
  
    instance.additionalProperties = new Map();
    const propsToCheck = Object.entries(obj).filter((([key,]) => {return !["type","displayName","email","additionalProperties"].includes(key);}));
    for (const [key, value] of propsToCheck) {
      instance.additionalProperties.set(key, value as any);
    }
    return instance;
  }
  public static theCodeGenSchema = {"type":"object","properties":{"type":{"const":"SimpleObject"},"displayName":{"type":"string","description":"Name of the user"},"email":{"type":"string","format":"email","description":"Email of the user"}}};
  public static validate(context?: {data: any, ajvValidatorFunction?: ValidateFunction, ajvInstance?: Ajv, ajvOptions?: AjvOptions}): { valid: boolean; errors?: ErrorObject[]; } {
    const {data, ajvValidatorFunction} = context ?? {};
    const parsedData = typeof data === 'string' ? JSON.parse(data) : data;
    const validate = ajvValidatorFunction ?? this.createValidator(context)
    return {
      valid: validate(parsedData),
      errors: validate.errors ?? undefined,
    };
  }
  public static createValidator(context?: {ajvInstance?: Ajv, ajvOptions?: AjvOptions}): ValidateFunction {
    const {ajvInstance} = {...context ?? {}, ajvInstance: new Ajv(context?.ajvOptions ?? {})};
    addFormats(ajvInstance);
    const validate = ajvInstance.compile(this.theCodeGenSchema);
    return validate;
  }

}
export { AnonymousSchema_1 };"
`;
